bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractZkLedgerManager.java

```diff
@@ -402,6 +402,7 @@ public void registerLedgerMetadataListener(long ledgerId, LedgerMetadataListener
                 }
             }
             synchronized (listenerSet) {
+                listenerSet = listeners.computeIfAbsent(ledgerId, k -> new HashSet<>());
```

Comment:
I think this should be before the synchronized block, not inside of it

Reply:
> I think this should be before the synchronized block, not inside of it

Maybe not, if outside, the race condition problem still exist.

Reply:
with this
```
            Set<LedgerMetadataListener> listenerSet = listeners.computeIfAbsent(ledgerId, k -> new HashSet<>());
            synchronized (listenerSet) {
                listenerSet = listeners.computeIfAbsent(ledgerId, k -> new HashSet<>());
                listenerSet.add(listener);
            }
```
it looks like the codes assume that between two calls of `listeners.computeIfAbsent` it may need to create a new instance (listeners can be modified/removed concurrently.) 

so you are getting (or creating) one instance to synchronize on, meanwhile, listeners change, now synchronize locks on one instance of listenerSet while modifying another.

so either the second call to computeIfAbsent is not needed or the locking needs to be done differently

Reply:
If we don't want second `computeIfAbsent` and still lock by ledgerId, we should define other object which is immutable. likes Map<String, Object> key: ledgerId, value: empty object. But it will introduce more object.
The simplest way is make register and unregister synchronize, but it will reduce concurrency.

In summary, I think second `computeIfAbsent` is the relatively good way.

Reply:
Let's use lock object (`private final Object lock = new Object();` is good enough); and replace other instance of `synchronized (listenerSet)` with `synchronized (lock) ...`
Otherwise I think it will just make the problem harder to reproduce

Reply:
It will reduce too many concurrency. If we listen too many zk nodes, when received event, also synchronized, it maybe make program slow.
see https://github.com/apache/bookkeeper/blob/ded05aa2218f1e98c650070e9d0152213c4d09fe/bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractZkLedgerManager.java#L125.